smart-contracts/test/mainnet/governance-proposal.js

```diff
@@ -0,0 +1,289 @@
+    voterWallet
+  )
+  const state = await governanceForVoter.state(proposalId)
+  if (state < 2) {
+    console.log(`${voterAddress} approves (votes 1)`)
+    await governanceForVoter.castVote(proposalId, 1)
+  }
+}
+
+/**
+ * Finishes the voting period!
+ * @param {*} proposalId
+ */
+const finishVotes = async (proposalId) => {
+  const gov = await new ethers.Contract(
+    governanceContractAddress,
+    UnlockProtocolGovernor.abi,
+    ethers.provider
+  )
+
+  const deadline = await gov.proposalDeadline(proposalId)
+  const currentBlock = await ethers.provider.getBlockNumber()
+  if (currentBlock < deadline) {
+    console.log('advancing to deadline at block #', deadline.toNumber())
+    await time.advanceBlockTo(deadline.toNumber())
+  }
+}
+
+/**
+ * Queues a proposal
+ * @param {*} proposerAddress
+ * @param {*} proposal
+ */
+const queueProposal = async (proposerAddress, proposalId, proposal) => {
+  const [targets, values, calldatas, description] = proposal
+  const descriptionHash = web3.utils.keccak256(description)
+  const voterWallet = await ethers.getSigner(proposerAddress)
+  const gov = await new ethers.Contract(
+    governanceContractAddress,
+    UnlockProtocolGovernor.abi,
+    voterWallet
+  )
+  const state = await gov.state(proposalId)
+  if (state === 5) {
+    const eta = await gov.proposalEta(proposalId)
+    return eta
+  }
+  const tx = await gov.queue(targets, values, calldatas, descriptionHash)
+  const { events } = await tx.wait()
+  const evt = events.find((v) => v.event === 'ProposalQueued')
+  const { eta } = evt.args
+  return eta.toNumber()
+}
+/**
+ * Executes a proposal
+ * @param {*} proposerAddress
+ * @param {*} proposal
+ */
+const executeProposal = async (executorAddress, proposal, timestamp) => {
+  const [targets, values, calldatas] = proposal
+  const descriptionHash = web3.utils.keccak256(proposal.slice(-1).find(Boolean))
+
+  const executerWallet = await ethers.getSigner(executorAddress)
+  const gov = await new ethers.Contract(
+    governanceContractAddress,
+    UnlockProtocolGovernor.abi,
+    executerWallet
+  )
+
+  const currentTime = (await time.latest()).toNumber()
+  if (currentTime < timestamp) {
+    await time.increaseTo(timestamp + 1)
+  }
+  await gov.execute(targets, values, calldatas, descriptionHash, {
+    gasLimit: '1000000',
+  })
+}
+
+/**
+ * Set state for pre-requisites to governance decision, should be customized for every proposal
+ */
+const prerequisites = async () => {
+  // Impersonate team wallet
+  await impersonate(teamWalletAddress)
+  const teamWallet = await ethers.getSigner(teamWalletAddress)
+  const udt = await new ethers.Contract(
+    udtContractAddress,
+    UnlockDiscountTokenV2.abi,
+    teamWallet
+  )
+
+  // Transfer 1 UDT from team's wallet to governance
+  await udt.transfer(timelockContractAddress, ethers.utils.parseUnits('1', 18))
+}
+
+/**
+ * This can be customized for every proposal!
+ * @returns
+ */
+const buildProposal = async () => {
```

Comment:
This logic should be exported by a file like `./proposals/change-lock-template.ts` (there it would be `./proposals/transfer-0.01-udt-to-0x12231.ts`)